### PR TITLE
Invisible command `for:` argument

### PR DESCRIPTION
**Pull request by tal5**
## Additions
- `invisible` command:

> - `for:<player>|...` argument - fakes the entity's visibility state for the specified players.
> - `state:reset` option - resets an entity's fake visibility state (for all players).
> - specifying no target will now default to the linked player (or NPC, if there isn't one).
> - Tab completion for the `state:` argument.
> - `#setInvisibleForPlayer(EntityTag target, PlayerTag player, boolean invisible)` - Sets an entity's invisibility state for a specific player.
> - `#isInvisible(Entity entity, UUID player, boolean fakeOnly)` - Returns whether an entity is invisible. if a player is given, will check whether an entity is invisible to that player as well. if `fakeOnly` is `true`, returns null if there's no fake visibility set for that player. 

- `EntityHelper#isInvisible` - returns weather an entity is invisible, counterpart of `EntityHelper#setInvisible`. Used for the `toggle` argument.

## Changes

- `InvisibleCommand#setInvisible` - flipped the boolean param, as it used to be weather the entity should be visible, not invisible.
- Using the `toggle` state with the `for:` argument will toggle the entity's invisibility for each player separately.
- Changed `DenizenNetworkManagerImpl#getModifiedMetadataFor` to set the `invisible` flag according to 
`InvisibleCommand.invisibleEntities`.
## Notes

- `EntityHelper#isInvisible` - is implemented in all supported versions as opposed to only recent ones, as it's a counterpart to `setInvisible` and they're used together for the `toggle` argument.
- `isInvisible/setInvisible` - checks / removes the invisibility potion effect for compact with old uses.

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java [hunk 1]
```diff
@@ -1,60 +1,79 @@
...
         isProcedural = false;
     }
 
     // <--[command]
     // @Name Invisible
-    // @Syntax invisible [<entity>] (state:true/false/toggle)
-    // @Required 1
-    // @Maximum 2
-    // @Short Makes an NPC or entity go invisible.
+    // @Syntax invisible [<entity>] (state:true/false/toggle/reset) (for:<player>|...)
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java]: Syntax meta needs to be updated if entity is now optional

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java [hunk 2]
```diff
@@ -1,60 +1,79 @@
...
     //
     // @Usage
-    // Use to makes the player invisible.
-    // - invisible <player> state:true
+    // Use to make the linked player (or NPC, if there isn't one) invisible.
+    // - invisible
     //
     // @Usage
-    // Use to make the attached NPC visible if previously invisible, and invisible if not
+    // Use to make the linked NPC visible if previously invisible, and invisible if not
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java]: should add the missing period at end of line

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java [hunk 3]
```diff
@@ -73,87 +96,130 @@ else if (!scriptEntry.hasObject("target")
                     && Utilities.entryHasNPC(scriptEntry)) {
                 scriptEntry.addObject("target", Utilities.getEntryNPC(scriptEntry).getDenizenEntity());
             }
-            else if (!scriptEntry.hasObject("target")
-                    && arg.matchesArgumentType(EntityTag.class)) {
-                scriptEntry.addObject("target", arg.asType(EntityTag.class));
-            }
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java]: This reordering likely will break the legacy `player` input support

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java [hunk 4]
```diff
@@ -73,87 +96,130 @@ else if (!scriptEntry.hasObject("target")
...
         }
     }
 
     @Override
     public void execute(ScriptEntry scriptEntry) {
-        ElementTag state = scriptEntry.getElement("state");
         EntityTag target = scriptEntry.getObjectTag("target");
-        if (scriptEntry.dbCallShouldDebug()) {
-            Debug.report(scriptEntry, getName(), state, target);
+        if (target == null || target.isFake) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java]: Why are you excluding fake entities here?

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java [hunk 5]
```diff
@@ -73,87 +96,130 @@ else if (!scriptEntry.hasObject("target")
...
+                }
+                else {
+                    for (PlayerTag player : forPlayers) {
+                        setInvisibleForPlayer(target, player, !isInvisible(target.getBukkitEntity(), player.getUUID(), false));
                     }
-                    else {
-                        setInvisible(target, target.getLivingEntity().hasPotionEffect(PotionEffectType.INVISIBILITY));
+                }
+                break;
+            case RESET:
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java:181]: Probably should handle `reset for:...`? (ie only reset the specified players)

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java (hunk 1 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java]: Changed to `(<entity>)`

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java (hunk 2 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java]: Added

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java (hunk 3 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java]: Reverted

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java (hunk 4 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java]: I... think it did that somewhere else and I moved it here? might've confused it with something - removed

Comment (mcmonkey4eva) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java]: The spot you're referring to isn't an exclusion, it's just a separate codepath dedicated to handling fake entities properly

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java (hunk 5 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java:181]: Added

**Pull request merged**